### PR TITLE
Add support for more types of layouts

### PR DESCRIPTION
**Pull request by JesusDeLaProg**
Multiple improvements have been made to support more types of layouts.

- During text matching : 
   - innerText is used instead of textContent in order to remove line breaks
   - All space-like characters are replaced with " "
- During the `<mark>` creation phase : 
   - Highlighting starts and ends exactly on selection start/end
   - All block elements are highlighted separately because `<mark>` elements cannot contain block elements
- processFragmentDirectives now returns an array of arrays of elements. This array contains the list of `<mark>` elements created for each text fragment.

### CODE REVIEW

File: src/text-fragment-utils.js [hunk 1]
```diff
@@ -192,33 +208,124 @@ const processTextFragmentDirective = (textFragment) => {
...
-        range.setEndAfter(endNode);
+      const startMarks = [];
+      let firstNode = startNode;
+      startMarks.push(
+        createMarkFor(firstNode, {
+          start: { node: firstNode, offset: startOffset },
+        }),
+      );
+      firstNode = startMarks[0];
+      while (firstNode.parentNode !== commonAncestor) {
```
Comment (tfmar) [inline src/text-fragment-utils.js:224]: Could you add a comment for this while loop explaining what the goal of walking the tree here is?

File: src/text-fragment-utils.js [hunk 2]
```diff
@@ -192,33 +208,124 @@ const processTextFragmentDirective = (textFragment) => {
...
+      // Need to create multiple <mark> elements to highlight the entire selection.
       const commonAncestor = range.commonAncestorContainer;
-      while (startNode.parentNode !== commonAncestor) {
-        startNode = startNode.parentNode;
-        range.setStartBefore(startNode);
-      }
-      while (endNode.parentNode !== commonAncestor) {
-        endNode = endNode.parentNode;
-        range.setEndAfter(endNode);
+      const startMarks = [];
```
Comment (tfmar) [inline src/text-fragment-utils.js:216]: Could you add comments to the startMarks/endMarks vars, and maybe a bit about why it's split up into start/inBetween/end this way?

File: src/text-fragment-utils.js [hunk 3]
```diff
@@ -192,33 +208,124 @@ const processTextFragmentDirective = (textFragment) => {
...
+    createdMarks.push(
+      createMarkFor(null, { startBefore: firstNode, endAfter: lastNode }),
+    );
+  }
+  return createdMarks.filter(Boolean);
+};
+
+/**
+ * Creates a mark and surround the wanted range in it. Will not create an element if the range is collapsed.
+ * @param {Node} baseNode
```
Comment (tfmar) [inline src/text-fragment-utils.js:299]: It seems like only one of `baseNode` or `options` is needed; can we document that?

File: src/text-fragment-utils.js [hunk 4]
```diff
@@ -192,33 +208,124 @@ const processTextFragmentDirective = (textFragment) => {
...
+/**
+ * Highlights the provided nodes entirely. Creates as few `<mark>` elements as possible.
+ * @param {Node[]} nodes
+ * @return {Element[]} mark elements created.
+ */
+const highlightNodes = (nodes) => {
+  if (!nodes || !nodes.length) return [];
+
+  const createdMarks = [];
+  let firstNode = nodes[0];
```
Comment (tfmar) [inline src/text-fragment-utils.js:275]: What are `firstNode` and `lastNode` tracking? It's a bit confusing that they change, and frequently equal each other.

File: src/text-fragment-utils.js [hunk 5]
```diff
@@ -192,33 +208,124 @@ const processTextFragmentDirective = (textFragment) => {
...
+  let lastNode = firstNode;
+  for (let i = 0; i < nodes.length; ++i) {
+    lastNode = nodes[i];
+    if (BLOCK_ELEMENTS.includes(lastNode.tagName)) {
+      createdMarks.push(
+        createMarkFor(null, { startBefore: firstNode, endBefore: lastNode }),
+      );
+      createdMarks.concat(highlightNodes(Array.from(lastNode.childNodes)));
```
Comment (tfmar) [inline src/text-fragment-utils.js:283]: Do you think we should set a max recursion depth? I wonder if performance would be noticeably bad if, say, you went to a really complex page and then picked the first and last elements as your textStart/textEnd.

File: src/text-fragment-utils.js [hunk 6]
```diff
@@ -238,13 +345,14 @@ export const scrollElementIntoView = (element) => {
...
  * Finds the DOM Node and the exact offset where a string starts or ends.
- * @param {Node} blockNode - Block element in which to search for a given text.
+ * @param {HTMLElement} blockNode - Block element in which to search for a given text.
  * @param {string} text - The text for which to find the position.
  * @param {boolean} start - Whether to return the an offset for the start or the text, or the end.
  * @return {[Node, number]} The DOM Node and the offset where the text starts or ends.
  */
 const findRangeNodeAndOffset = (blockNode, text, start) => {
-  let offset = blockNode.textContent.indexOf(text) + (start ? 0 : text.length);
+  const fullText = blockNode.innerText.replace(/\s/g, ' ');
```
Comment (tfmar) [inline src/text-fragment-utils.js:354]: I wonder if we can get away without this. Shouldn't the URL encode the same kind of space character as the node's text?

File: src/text-fragment-utils.js [hunk 7]
```diff
@@ -284,27 +392,47 @@ const findText = (text) => {
...
     return [];
   }
-  const body = document.body;
-  const treeWalker = document.createTreeWalker(body, NodeFilter.SHOW_ELEMENT, {
-    acceptNode: (node) => {
-      if (!BLOCK_ELEMENTS.includes(node.tagName)) {
-        return NodeFilter.FILTER_REJECT;
-      }
+
+  // Accept only block elements that do not contain other block elements and
```
Comment (tfmar) [inline src/text-fragment-utils.js:396]: Isn't it possible that a node would contain the text (but not in a child) and a block element child?

```
  <div>
    This is my matching text 
    <p> and this is an irrelevant block element </p> 
  </div>
```

File: src/text-fragment-utils.js [hunk 8]
```diff
@@ -22,6 +22,22 @@
  * @property {string} suffix
  */
 
+/**
+ * @typedef NodeAndOffset
+ * @property {Node} node
+ * @property {number} offset
```
[the remark this answers is not in the document]
Comment (JesusDeLaProg) [reply src/text-fragment-utils.js]: Removed those typedefs because they became unused after refactoring.

File: src/text-fragment-utils.js (hunk 7 above)
Comment (JesusDeLaProg) [reply src/text-fragment-utils.js]: This new version should work in this case and many more!

**Pull request merged**